contents/c++/rvalue.md

```diff
@@ -0,0 +1,169 @@
+* Indeterminant behaviour in a multi-threaded environment. Since each thread will all be pointing to the same object, this will lead to unnecessary data races and indeterminant behaviour as these same objects are modified concurrently in each thread. This in turn requires extra work to synchronize the values between objects, which also slows down the program. More details of this can be found in the resources [Java Concurrency](java/JavaConcurrency.md) and [Java Synchronization](java/JavaSynchronization.md).
+* Unintended behaviour. If users are not careful, they can modify the values in the object using one pointer or reference, which will then be reflected by another pointer to the same object, which the user may not have intended. A good example of this problem can be found in this [article](http://www.drdobbs.com/cpp/optimization-calling-by-value-or-by-refe/232400151).
+
+## Move Semantics
+From the summary above, there are many benefits for using Value Semantics. However, it's one major flaw is that a copy of the object will always be created, which can computationally expensive if this object is of a large size. However, Reference Semantics may also not be preferred due to its different problems as discussed above. As such, in order to continue gaining the benefits of Value Semantics while overcoming it's major flaw, `C++11` introduced a new mode to users, Move Semantics.
+
+### rvalue and lvalue references
+To understand how Move Semantics work in C++, it is important to distinguish between rvalue and lvalue references.
+```
+lvalue = rvalue
+```
+From the line above, lvalue (left value) basically refers to values that are addressable, whileas rvalue (right value) are temporary objects or values which are used only on the right side of an assignment expression. More details and classification of these 2 values can be found [here](http://www.bogotobogo.com/cplusplus/C11/4_C11_Rvalue_Lvalue.php).
+
+### Using rvalue references
+Rvalue references allows us to distinguish an lvalue from an rvalue. In C++11, we can declare an rvalue reference using the `&&` operator:
+```cpp
+int &&rvalue = 55;
+```
+
+We can also convert an lvalue to a rvalue using the `std::move` function:
+```cpp
+int lvalue = 99;
+int &&rvalue2 = std::move(lvalue);
+```
+
+We can also do [function overloading](https://en.wikibooks.org/wiki/Computer_Programming/Function_overloading)
+in order to determine whether the parameters given are lvalues or rvalues:
+
+```cpp
+void print(int& lvalue) { // takes an lvalue
+    std::cout << "lvalue method used";
+}
+
+void print(int&& rvalue) { // takes a rvalue
+    std::cout << "rvalue method used";
+}
+
+int main() {
+    int x = 5;
+    print(x); // lvalue method used
+    print(10); // rvalue method used.
+}
+```
+
+Using of these functions overload with rvalue parameters, which take on temporary objects, helps in writing more efficient program using Move Semantics!
```

Comment:
Can be more concise here. Also, there are some typos here as well. Could re-written as:
**Usage** of function overloading with rvalue parameters, help in writing more efficient program`s` using Move Semantics!